### PR TITLE
fix(parser) reject oauth2s without name

### LINKED ISSUE #1505: Oauth2 credentials without Name field result in segfault
Oauth2 credentials with no `name` field are [accepted by KIC's credential validation](https://github.com/Kong/kubernetes-ingress-controller/blob/1.3.1/pkg/kongstate/credentials.go#L124-L134). They are not accepted by decK's libraries, and their omission [will result in a segfault](https://github.com/Kong/deck/blob/v1.7.0/state/types.go#L929). The field is required in the [Kong plugin DAO](https://github.com/Kong/kong/blob/2.4.1/kong/plugins/oauth2/daos.lua#L33) and we should ensure that it is present in KIC.

**Kong Ingress controller version**
 ingress-controller - 1.2

**Kong or Kong Enterprise version**
Any.

### What happened
KIC crashed.

### Expected behavior
KIC does not crash.

### Steps To Reproduce

1. Create an Oauth2 credential with no name field.
2. Watch KIC crash.

See original report in https://github.com/Kong/kong/issues/7517 for additional details.

To fix, we need to add a new clause to `NewOauth2Credential()` which returns an error if `res.Name` is not set, similar to the existing clause for `res.ClientID`.

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
Require a Name field in oauth2 credentials.

Upstream code requires this field. Omitting it currently produces a segfault. We would rather reject these credentials and log a warning rather than segfault.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 
Fix #1505

**Special notes for your reviewer**:
This is going to main because it's a bugfix. Our new changelog with PR process works well for stuff in `next`, where we know there will someday be a new minor release, but is a bit ambiguous for `main`, where we may release a patch release or may simply roll fixes into the next minor.

How should we properly update the staged changelog for patch fixes that go directly to main, so that the releaser knows to include them in either the patch release (if one happens) or the minor release (if that comes first)? Stuff going to main also means merge conflicts in the changelog, though not very complicated ones.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: pkg/kongstate/credentials.go [hunk 1]
```diff
@@ -130,6 +130,9 @@ func NewOauth2Credential(config interface{}) (*Oauth2Credential, error) {
 	if res.ClientID == nil {
 		return nil, fmt.Errorf("oauth2 is invalid: no client_id")
 	}
+	if res.Name == nil {
```
Comment (ccfishk) [inline pkg/kongstate/credentials.go:133]: Do we apply or need to apply any name convention ? 

Fo example, k8s resource generically follow this name convention https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names

File: pkg/kongstate/consumer_test.go [hunk 2]
```diff
@@ -310,10 +310,25 @@ func TestConsumer_SetCredential(t *testing.T) {
...
 			name: "oauth2 without client_id",
 			args: args{
-				credType:   "oauth2",
-				consumer:   &Consumer{Consumer: kong.Consumer{Username: &username}},
-				credConfig: map[string]string{},
-				version:    standardVersion,
+				credType: "oauth2",
+				consumer: &Consumer{Consumer: kong.Consumer{Username: &username}},
+				credConfig: map[string]interface{}{
+					"client_id": "bar",
```
Comment (ccfishk) [inline pkg/kongstate/consumer_test.go:316]: The TC name "oauth2 without client_id", but with client_id configured. Which way we should proceed ?

File: pkg/kongstate/consumer_test.go [hunk 3]
```diff
@@ -310,10 +310,25 @@ func TestConsumer_SetCredential(t *testing.T) {
...
+				},
+				version: standardVersion,
+			},
+			result:  &Consumer{Consumer: kong.Consumer{Username: &username}},
+			wantErr: true,
```
Comment (ccfishk) [inline pkg/kongstate/consumer_test.go:321]: Do we expect an error for this test case ?

File: pkg/kongstate/credentials.go (hunk 1 above)
Comment (rainest) [reply pkg/kongstate/credentials.go:133]: No. These are Kong fields inside the K8S object and don't have restrictions beyond being strings.

File: pkg/kongstate/consumer_test.go (hunk 2 above)
Comment (rainest) [reply pkg/kongstate/consumer_test.go:316]: Copy-pasted and unchanged test names are our favorite passtime: e4b3e45d

File: pkg/kongstate/consumer_test.go (hunk 3 above)
Comment (rainest) [reply pkg/kongstate/consumer_test.go:321]: Yes, it lacks a name and the new validation should drop it with an error return.

**Review by ccfishk — APPROVED** [review]
lgtm.

File: pkg/kongstate/consumer_test.go (hunk 2 above)
Comment (ccfishk) [reply pkg/kongstate/consumer_test.go:316]: 😆

**Pull request merged**